### PR TITLE
v0.13.0 Introducing the Rolodex. 

### PR DESCRIPTION
**Pull request by daveshanley**
The rolodex is the solution `libopenapi` has needed for some time. It reorganizes how the index works, how references are composed and how things are looked up when building the model. 

The rolodex creates virtual filesystems for exploded specifications that are scattered across multiple local or remote locations. The rolodex can contain multiple local or remote filesystems for documents, and can lookup anything
via its fully qualified location.

Custom filesystems can be loaded in, instead of the local and remote ones provided by `libopenapi` as the rolodex uses `fs.FS` as an interface, so any custom implementation can be used. 

This is a large breaking change for how the index operates, there are breaking changes to all low-level models, mainly with the introduction of a `context.Context` parameter. This is used by the rolodex to know where to resolve references from, when deeply nested recursive reference resolution. 

The index no-longer operates the way it used to and how indexes are used by the rolodex is also new. 

There should be minimal configuration changes at the top-level API for documents, and no changes to the high-level model.

### CODE REVIEW

**Comment by daveshanley** [comment]
Looks like the lovely single core runners are choking on locks.

**Comment by TristanSpeakEasy** [comment]
eagerly awaiting this

**Comment by daveshanley** [comment]
slogging through bumping test coverage, a lot of this was built in the hot path.

**Comment by daveshanley** [comment]
Today, this damn PR will go green.

**Comment by TristanSpeakEasy** [comment]
> Today, this damn PR will go green.

 I have been watching all the commits with great interest haha

**Comment by daveshanley** [comment]
Adding some docs and cleaning things up and will be ready for review.

**Comment by daveshanley** [comment]
For end users of the high-level model. You should notice no real changes. However this PR is loaded with breaking changes, I had to open up the heart of the library and add new bones to the skeleton.

File: index/spec_index.go [hunk 1]
```diff
@@ -14,31 +14,43 @@ package index
...
 // NewSpecIndexWithConfig will create a new index of an OpenAPI or Swagger spec. It uses the same logic as NewSpecIndex
 // except it sets a base URL for resolving relative references, except it also allows for granular control over
 // how the index is set up.
 func NewSpecIndexWithConfig(rootNode *yaml.Node, config *SpecIndexConfig) *SpecIndex {
 	index := new(SpecIndex)
-	if config != nil && config.seenRemoteSources == nil {
-		config.seenRemoteSources = &syncmap.Map{}
-	}
-	config.remoteLock = &sync.Mutex{}
+	//if config != nil && config.seenRemoteSources == nil {
```
Comment (TristanSpeakEasy) [inline index/spec_index.go:33]: old code or todo?

File: index/spec_index.go [hunk 2]
```diff
@@ -89,17 +100,20 @@ func createNewIndex(rootNode *yaml.Node, index *SpecIndex, avoidBuildOut bool) *
 	}
 
 	// do a copy!
-	index.config.seenRemoteSources.Range(func(k, v any) bool {
-		index.seenRemoteSources[k.(string)] = v.(*yaml.Node)
-		return true
-	})
+	//index.config.seenRemoteSources.Range(func(k, v any) bool {
```
Comment (TristanSpeakEasy) [inline index/spec_index.go:103]: ditto

File: index/spec_index.go [hunk 3]
```diff
@@ -423,10 +446,10 @@ func (index *SpecIndex) GetAllOperationsServers() map[string]map[string][]*Refer
 	return index.opServersRefs
 }
 
-// GetAllExternalIndexes will return all indexes for external documents
-func (index *SpecIndex) GetAllExternalIndexes() map[string]*SpecIndex {
-	return index.externalSpecIndex
-}
+//// GetAllExternalIndexes will return all indexes for external documents
```
Comment (TristanSpeakEasy) [inline index/spec_index.go:449]: ditto

File: index/spec_index.go [hunk 4]
```diff
@@ -957,20 +984,24 @@ func (index *SpecIndex) GetOperationCount() int {
 							Path:       fmt.Sprintf("$.paths.%s.%s", p.Value, m.Value),
 							ParentNode: m,
 						}
-						index.pathRefsLock.Lock()
-						if index.pathRefs[p.Value] == nil {
-							index.pathRefs[p.Value] = make(map[string]*Reference)
+						//index.pathRefsLock.Lock()
```
Comment (TristanSpeakEasy) [inline index/spec_index.go:987]: more old code?

File: index/spec_index.go [hunk 5]
```diff
@@ -957,20 +984,24 @@ func (index *SpecIndex) GetOperationCount() int {
...
+						if locatedPathRefs[p.Value] == nil {
+							locatedPathRefs[p.Value] = make(map[string]*Reference)
 						}
-						index.pathRefs[p.Value][ref.Name] = ref
-						index.pathRefsLock.Unlock()
+						locatedPathRefs[p.Value][ref.Name] = ref
+						//index.pathRefsLock.Unlock()
```
Comment (TristanSpeakEasy) [inline index/spec_index.go:992]: ditto

**Review by TristanSpeakEasy — APPROVED** [review]
Looking good to me, nothing too scary in here from a migration perspective. Going to start a branch in our product and give the upgrade a try and see if we find any early issues

File: index/spec_index.go (hunk 1 above)
Comment (daveshanley) [reply index/spec_index.go:33]: old code, will remove.

File: index/spec_index.go (hunk 2 above)
Comment (daveshanley) [reply index/spec_index.go:103]: old code, needs to go.

File: index/spec_index.go (hunk 3 above)
Comment (daveshanley) [reply index/spec_index.go:449]: same

File: index/spec_index.go (hunk 4 above)
Comment (daveshanley) [reply index/spec_index.go:987]: yep

File: index/spec_index.go (hunk 5 above)
Comment (daveshanley) [reply index/spec_index.go:992]: same same..

**Comment by daveshanley** [comment]
It's time to merge.

**Pull request merged**